### DISCUSSION TITLE
JOB_DATA data type as VARCHAR(MAX)

### DISCUSSION THREAD

**Discussion by troygould** [quartznet/quartznet#676] [General] [answered]
Just a question.  
I see that in Quartz.NET version 3.0 that the JOB_DATA column has been changed from IMAGE to VARBINARY(MAX). Is there any issue with me changing this to VARCHAR(MAX) if I'm using the following settings?

            "quartz.jobStore.useProperties" = "true";
            "quartz.serializer.type" = "json";

I've done it, and have been able to successfully add/remove and run jobs with scheduled triggers with no issues so far. I just don't know if there is something lurking in the framework that I'm not aware of that will eventually cause an issue.

We don't store a ton of Jobs. So, space isn't an issue for us. However, readability of the raw data is arguably easier with VARCHAR versus VARBINARY.

Thanks in advance.

**Comment by lahma** [comment, marked as answer]
The VARBINARY comes from the need to serialize binary data, all the way from the beginning of the project. If you never ever need to serialize anything else than JSON, this probbaly works. I'm just not giving any guarantees as tests are run against 'standard' database and if ever needed, migration scripts would not account for your non-standard column type.

So if you have tests in place to validate everything works, I guess it doesn't hurt - just not supported 😃

**Reply by vladirimia** [reply]
I can not believe this is the solution for a similar problem I had, and it's 2025.
The context: .NET project using Quartz 3.14.
IF I set the connection string
```
options.UseSqlServer(sqlServer =>
{
    sqlServer.TablePrefix = "dbo.QRTZ_";
    sqlServer.ConnectionString = "myConnectionString";
});
```
everything works perfect with the VARBINARY(MAX) column in QRTZ_JOB_DETAILS and QRTZ_TRIGGERS.

Now, I needed to use a connection provider because my database is in Azure and each connection has an access token attached.
I created the connection provider and I have this configuration for Quartz:
```
services.AddQuartz(q =>
{
    q.SchedulerId = "AppScheduler";

    q.UsePersistentStore(options =>
    {
        options.UseNewtonsoftJsonSerializer();

        options.UseSqlServer(sqlServer =>
        {
            sqlServer.TablePrefix = "dbo.QRTZ_";
        });
        options.PerformSchemaValidation = false;
        
        options.SetProperty("quartz.jobStore.driverDelegateType", "Quartz.Impl.AdoJobStore.SqlServerDelegate, Quartz");
        options.SetProperty("quartz.jobStore.dataSource", "default");
        options.SetProperty("quartz.dataSource.default.connectionProvider.type", typeof(QuartzConnectionProvider).AssemblyQualifiedName);
    });
});
```
With this configuration, _scheduler.AddJob(myJob) works ok, it inserts the data with VARBINARY(MAX) as it should. The generated sql uses the correct data type.
Now, when I add a trigger for that job, the operation fails because it tries to use an varchar(max) parameter in the generated SQL instead of varbinary(max).
I haven't found any workaround, or any explanation on what happens with the trigger and why inserting a trigger is different that inserting a job in the database.

I will dig into the code, maybe I can find the implementation, but it's really... really weird.
